### PR TITLE
Protocol tests issue 33

### PR DESCRIPTION
**Pull request by peterrsongg**
## Description

This PR fixes DOTNET-7357 issue 33. `Issue 33 Issue unmarshalling Xml Responses due to improper handling of "LocationName"`

* (Review manual changes) commit 6f05ab0d77b - Fixes for issue 33.
* (Reference generated changes) commit 443a6204a24 - Generated changes for issue 33.

**MERGE into `protocol-tests` feature branch only**

Builds on prior PR: https://github.com/aws/aws-sdk-net/pull/3266
Merge order: 112 (I'm going to start mine from 100 and increment to keep separate from Bo's)
Ticket: DOTNET-7357

## Testing

N/A - the test cases are the protocol tests. A final dry-run will be done on the `protocol-tests` feature branch once everything is merged into the feature branch.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt [hunk 1]
```diff
@@ -256,4 +257,29 @@ namespace <#=this.Config.Namespace #>.Model.Internal.MarshallTransformations
     this.AddResponseSingletonMethod();
 #>
     }
-}
\ No newline at end of file
+}
+<#+
+    /* The rules for determining the marshallName for xml services is different than json services
+        *  1. Use the member locationName if the value doesn't match the member's name
```
Comment (boblodgett) [inline generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt:263]: nit: extra `* `

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt [hunk 2]
```diff
@@ -256,4 +257,29 @@ namespace <#=this.Config.Namespace #>.Model.Internal.MarshallTransformations
...
+    {
+        var locationName = member.data[ServiceModel.LocationNameKey];
+        if(locationName == null)
+            return member.MarshallName;
+        if (!string.Equals(locationName.ToString(), member.ModeledName, StringComparison.Ordinal))
+        {
+            return locationName.ToString();
+        }
+        var memberTarget = member.Shape.data[ServiceModel.LocationNameKey];
+        if(member.Shape.data != null && memberTarget != null)
```
Comment (boblodgett) [inline generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt:277]: `member.Shape.data != null` doesn't seem needed as you would have referenced a location in a null json object previously and that would have failed. Does the check need to be moved or removed?

Comment (peterrsongg) [reply generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt]: you're right, Member.Shape.data would never actually return null I removed it

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt [hunk 3]
```diff
@@ -3,6 +3,7 @@
 <#@ import namespace="System.Linq" #>
 <#@ import namespace="System.Text" #>
 <#@ import namespace="System.Collections.Generic" #>
+<#@ import namespace="Json.LitJson" #>
```
[the remark this answers is not in the document]
Comment (peterrsongg) [reply generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt]: I forgot to remove this after experimenting. removed.

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt [hunk 4]
```diff
@@ -256,4 +257,29 @@ namespace <#=this.Config.Namespace #>.Model.Internal.MarshallTransformations
...
+        if (!string.Equals(locationName.ToString(), member.ModeledName, StringComparison.Ordinal))
+        {
+            return locationName.ToString();
+        }
+        var memberTarget = member.Shape.data[ServiceModel.LocationNameKey];
+        if(member.Shape.data != null && memberTarget != null)
+        {
+            return memberTarget.ToString();
+        }
+        else{
```
[the remark this answers is not in the document]
Comment (peterrsongg) [reply generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlResponseUnmarshaller.tt]: thanks for the suggestions. I applied the changes

**Comment by peterrsongg** [comment]
There are some additional generated changes that snuck in there. These are files that I didn't push in a different commit. You can ignore those. Once we merge all these in we are going to have to run the generator again anyways.

**Review by boblodgett — APPROVED** [review]

**Review by dscpinheiro — APPROVED** [review]

**Pull request merged**